### PR TITLE
[runtime] Add GC Unsafe transitions to some MONO_API functions

### PR DESCRIPTION
**Pull request by lambdageek**
For Hybrid suspend #6921 to be seamless for embedders, we want each Mono function that can be called from GC Safe code to perform its own transition to GC Unsafe.

This PR adds the transitions to about a third of the functions in the list in this comment: https://github.com/mono/mono/issues/6921#issuecomment-390757584

### CODE REVIEW

**Comment by lambdageek** [comment]
Linux x64 FullAOT failure seems unrelated - what's `MONO_DEBUG=test-tailcall-require` ?

```
MONO_DEBUG=test-tailcall-require MONO_PATH=/mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot/mcs/class/lib/build ../../runtime/mono-wrapper --aot-path=/mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot/mcs/class/lib/build --debug ./test-runner.exe  --config tests-config --runtime "../../runtime/mono-wrapper" --mono-path "/mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot/mcs/class/lib/testing_aot_full" --runtime-args "--runtime=mobile --full-aot"  --runtime-args ""  --verbose -j a  --testsuite-name "runtime-tailcall" --timeout 300 --disabled "delegate-async-exception.exe bug-348522.2.exe bug-459094.exe delegate-invoke.exe bug-Xamarin-5278.exe appdomain-marshalbyref-assemblyload.exe abort-try-holes.exe threads-init.exe dim-constrainedcall.exe gptail1.exe itaili1.exe sirtail1.exe srtail1.exe stail1.exe tail1.exe taili1.exe vtail1.exe tailcall-member-function-in-valuetype.exe tailcall-rgctxb.exe tailcall/coreclr/JIT/Directed/IL/PInvokeTail/tailwinapi.exe tailcall/coreclr/JIT/Directed/IL/Tailcall/Jittailcall2.exe tailcall/coreclr/JIT/Directed/IL/mutualrecur-tailcall/MutualRecur-TailCall.exe tailcall/coreclr/JIT/Directed/coverage/importer/Desktop/badtailcall.exe tailcall/coreclr/JIT/Directed/coverage/importer/badtailcall.exe tailcall/coreclr/JIT/Directed/pinvoke/tail.exe tailcall/coreclr/JIT/IL_Conformance/Old/Base/tailcall.exe tailcall/coreclr/JIT/Methodical/Invoke/fptr/recurse_tail_call.exe tailcall/coreclr/JIT/Methodical/Invoke/fptr/recurse_tail_calli.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_array.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_array_nz.exe tailcall/coreclr/JIT/Methodical/tailcall/gcval_nested.exe tailcall/coreclr/JIT/Methodical/tailcall/pointer_i.exe tailcall/coreclr/JIT/Methodical/tailcall/pointer.exe tailcall/coreclr/JIT/Methodical/tailcall/reference_i.exe tailcall/coreclr/JIT/Methodical/tailcall/test_implicit.exe tailcall/coreclr/JIT/Methodical/tailcall/test_mutual_rec.exe tailcall/coreclr/JIT/Methodical/tailcall/test_switch.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/delegateTail.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/hijacking.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/smallFrame.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/tailcall_AV.exe tailcall/coreclr/JIT/Regression/VS-ia64-JIT/V1.2-M02/b102844/tailcallcalli.exe tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyWithPrefix.exe tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyTransparentLibraryWithPrefix.exe tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyVerifiableLibraryWithPrefix.exe tailcall/coreclr/JIT/opt/ETW/TailCallCases.exe tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyWithPrefix.exe tailcall/coreclr/JIT/Methodical/Boxing/misc/tailjump.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_gc.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_inst.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_value.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_virt.exe tailcall/coreclr/JIT/Methodical/tailcall/gcval.exe tailcall/coreclr/JIT/Methodical/tailcall/gcval_sideeffect.exe tailcall/coreclr/JIT/Methodical/tailcall/test_3b.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/delegateParamCallTarget.exe  tailcall/coreclr/JIT/Directed/IL/PInvokeTail/tailwinapi.exe tailcall/coreclr/JIT/Directed/pinvoke/tail.exe tailcall/coreclr/JIT/Methodical/Invoke/fptr/recurse_tail_call.exe tailcall/coreclr/JIT/Methodical/Invoke/fptr/recurse_tail_calli.exe tailcall/coreclr/JIT/Methodical/tailcall/Desktop/thread-race.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_array_nz.exe tailcall/coreclr/JIT/Methodical/tailcall/pointer.exe tailcall/coreclr/JIT/Methodical/tailcall/reference_i.exe tailcall/coreclr/JIT/Methodical/tailcall/test_mutual_rec.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/delegateTail.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/tailcall_AV.exe tailcall/coreclr/JIT/Regression/VS-ia64-JIT/V1.2-M02/b102844/tailcallcalli.exe tailcall-member-function-in-valuetype.exe tailcall-rgctxb.exe  finally_guard.exe          tailcall/coreclr/JIT/Directed/IL/PInvokeTail/tailwinapi.exe tailcall/coreclr/JIT/Directed/IL/Tailcall/Jittailcall2.exe tailcall/coreclr/JIT/Directed/IL/mutualrecur-tailcall/MutualRecur-TailCall.exe tailcall/coreclr/JIT/Directed/coverage/importer/Desktop/badtailcall.exe tailcall/coreclr/JIT/Directed/coverage/importer/badtailcall.exe tailcall/coreclr/JIT/Directed/pinvoke/tail.exe tailcall/coreclr/JIT/IL_Conformance/Old/Base/tailcall.exe tailcall/coreclr/JIT/Methodical/Invoke/fptr/recurse_tail_call.exe tailcall/coreclr/JIT/Methodical/Invoke/fptr/recurse_tail_calli.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_array.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_array_nz.exe tailcall/coreclr/JIT/Methodical/tailcall/gcval_nested.exe tailcall/coreclr/JIT/Methodical/tailcall/pointer_i.exe tailcall/coreclr/JIT/Methodical/tailcall/pointer.exe tailcall/coreclr/JIT/Methodical/tailcall/reference_i.exe tailcall/coreclr/JIT/Methodical/tailcall/test_implicit.exe tailcall/coreclr/JIT/Methodical/tailcall/test_mutual_rec.exe tailcall/coreclr/JIT/Methodical/tailcall/test_switch.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/delegateTail.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/hijacking.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/smallFrame.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/tailcall_AV.exe tailcall/coreclr/JIT/Regression/VS-ia64-JIT/V1.2-M02/b102844/tailcallcalli.exe tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyWithPrefix.exe tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyTransparentLibraryWithPrefix.exe tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyVerifiableLibraryWithPrefix.exe tailcall/coreclr/JIT/opt/ETW/TailCallCases.exe tailcall/coreclr/JIT/opt/Tailcall/TailcallVerifyWithPrefix.exe tailcall/coreclr/JIT/Methodical/Boxing/misc/tailjump.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_gc.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_inst.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_value.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_virt.exe tailcall/coreclr/JIT/Methodical/tailcall/gcval.exe tailcall/coreclr/JIT/Methodical/tailcall/gcval_sideeffect.exe tailcall/coreclr/JIT/Methodical/tailcall/test_3b.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/delegateParamCallTarget.exe  tailcall/coreclr/JIT/Directed/IL/PInvokeTail/tailwinapi.exe tailcall/coreclr/JIT/Directed/pinvoke/tail.exe tailcall/coreclr/JIT/Methodical/Invoke/fptr/recurse_tail_call.exe tailcall/coreclr/JIT/Methodical/Invoke/fptr/recurse_tail_calli.exe tailcall/coreclr/JIT/Methodical/tailcall/Desktop/thread-race.exe tailcall/coreclr/JIT/Methodical/tailcall/deep_array_nz.exe tailcall/coreclr/JIT/Methodical/tailcall/pointer.exe tailcall/coreclr/JIT/Methodical/tailcall/reference_i.exe tailcall/coreclr/JIT/Methodical/tailcall/test_mutual_rec.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/delegateTail.exe tailcall/coreclr/JIT/Methodical/tailcall_v4/tailcall_AV.exe tailcall/coreclr/JIT/Regression/VS-ia64-JIT/V1.2-M02/b102844/tailcallcalli.exe tailcall-member-function-in-valuetype.exe tailcall-rgctxb.exe typeload-unaligned.exe field-access.exe invalid_generic_instantiation.exe bug-481403.exe array_ldelema.exe array_load_exception.exe bug445361.exe generic-type-load-exception.2.exe invalid-token.exe call_missing_method.exe call_missing_class.exe ldfld_missing_field.exe ldfld_missing_class.exe vt-sync-method.exe resolve_method_bug.2.exe resolve_field_bug.2.exe resolve_type_bug.2.exe bug-81691.exe bug-327438.2.exe newobj-abstract.exe context-static.exe bug-415577.exe generic-marshalbyref.2.exe unhandled-exception-7.exe marshal.exe marshal2.exe marshal6.exe marshal7.exe marshal8.exe pinvoke-2.2.exe pinvoke3.exe thunks.exe bug-58782-plain-throw.exe bug-58782-capture-and-throw.exe assembly-load-remap.exe remoting4.exe remoting1.exe remoting2.exe remoting3.exe remoting5.exe appdomain.exe appdomain-client.exe appdomain-unload.exe appdomain-async-invoke.exe appdomain-thread-abort.exe appdomain1.exe appdomain2.exe appdomain-exit.exe appdomain-unload-asmload.exe appdomain-unload-callback.exe appdomain-unload-doesnot-raise-pending-events.exe unload-appdomain-on-shutdown.exe appdomain-marshalbyref-assemblyload.exe assemblyresolve_event2.2.exe assemblyresolve_event6.exe bug-544446.exe bug-36848.exe generic-marshalbyref.2.exe stackframes-async.2.exe transparentproxy.exe bug-48015.exe delegate9.exe marshal-valuetypes.exe xdomain-threads.exe monitor.exe generic-xdomain.2.exe threadpool-exceptions7.exe cross-domain.exe generic-unloading.2.exe appdomain-threadpool-unload.exe bug-47295.exe loader.exe pinvoke2.exe generic-type-builder.2.exe dynamic-generic-size.exe cominterop.exe dynamic-method-access.2.exe dynamic-method-finalize.2.exe dynamic-method-stack-traces.exe generic_type_definition.2.exe bug-333798-tb.2.exe bug-335131.2.exe bug-322722_patch_bx.2.exe bug-322722_dyn_method_throw.2.exe bug-389886-2.exe bug-349190.2.exe bug-389886-sre-generic-interface-instances.exe custom-modifiers-inheritance.exe bug-462592.exe bug-575941.exe bug-389886-3.exe constant-division.exe dynamic-method-resurrection.exe assembly_append_ordering.exe assemblyresolve_event5.exe bug-80307.exe namedmutex-destroy-race.exe thread6.exe reference-loader.exe assembly-load-bytes.exe assembly-load-bytes-bindingredirect.exe constraints-load.exe bug-515884.exe generic-type-builder.2.exe sgen-domain-unload.exe sgen-domain-unload-2.exe appdomain-loader.exe assemblyresolve_event3.exe appdomain-serialize-exception.exe delegate15.exe weak-fields.exe tailcall/fsharp-deeptail.exe tailcall-interface-conservestack.exe   main-returns-background-resetabort.exe main-returns-background-abort-resetabort.exe assemblyresolve_event3.exe delegate2.exe finally_guard.exe generic-xdomain.2.exe  " tailcall/coreclr/JIT/CodeGenBringUpTests/RecursiveTailCall.exe tailcall/coreclr/JIT/opt/FastTailCall/FastTailCallCandidates.exe tailcall/coreclr/JIT/opt/FastTailCall/FastTailCallInlining.exe tailcall/coreclr/JIT/opt/FastTailCall/GitHubIssue12479.exe tailcall/coreclr/JIT/opt/FastTailCall/StackFixup.exe tailcall/coreclr/JIT/opt/FastTailCall/StructPassingSimple.exe tailcall/coreclr/JIT/Directed/IL/Tailcall/JitTailcall1.exe tailcall/coreclr/JIT/Directed/tailcall/tailcall.exe tailcall/coreclr/JIT/Methodical/Boxing/boxunbox/tailcall.exe tailcall/coreclr/JIT/Methodical/Invoke/SEH/catchfault_tail.exe tailcall/coreclr/JIT/Methodical/Invoke/SEH/catchfinally_tail.exe tailcall/coreclr/JIT/Methodical/nonvirtualcall/tailcall.exe tailcall/coreclr/JIT/Methodical/tailcall/compat_obj.exe tailcall/coreclr/JIT/Methodical/tailcall/compat_v.exe tailcall/coreclr/JIT/Methodical/tailcall/recurse_ep.exe tailcall/coreclr/JIT/Methodical/tailcall/recurse_ep_void.exe tailcall/coreclr/JIT/Methodical/tailcall/test_2a.exe tailcall/coreclr/JIT/Methodical/tailcall/test_2b.exe tailcall/coreclr/JIT/Methodical/tailcall/test_2c.exe tailcall/coreclr/JIT/Methodical/tailcall/test_virt.exe tailcall/coreclr/JIT/Methodical/tailcall/test_void.exe
Invalid option for the MONO_DEBUG env variable: test-tailcall-require
Available options: 'handle-sigint', 'keep-delegates', 'reverse-pinvoke-exceptions', 'collect-pagefault-stats', 'break-on-unverified', 'no-gdb-backtrace', 'suspend-on-native-crash', 'suspend-on-sigsegv', 'suspend-on-exception', 'suspend-on-unhandled', 'dont-free-domains', 'dyn-runtime-invoke', 'gdb', 'explicit-null-checks', 'gen-seq-points', 'no-compact-seq-points', 'single-imm-size', 'init-stacks', 'casts', 'soft-breakpoints', 'check-pinvoke-callconv', 'use-fallback-tls', 'debug-domain-unload', 'partial-sharing', 'align-small-structs', 'native-debugger-break', 'thread-dump-dir=DIR'
make[4]: *** [test-tailcall] Error 1
Makefile:2016: recipe for target 'test-tailcall' failed
```

File: mono/mini/mini-runtime.c [hunk 1]
```diff
@@ -3513,10 +3513,14 @@ mono_get_delegate_virtual_invoke_impl (MonoMethodSignature *sig, MonoMethod *met
...
 	// Empty string is ok as consequence of appending ",foo"
 	// without first checking for empty.
-	if (*option == 0)
-		return TRUE;
+	if (*option == 0) {
+		result = TRUE;
+		goto done;
+	}
 
 	if (!strcmp (option, "handle-sigint"))
```
Comment (jaykrell) [inline mono/mini/mini-runtime.c]: Could invert check and save indentation.

File: mono/metadata/assembly.c [hunk 2]
```diff
@@ -4430,7 +4458,11 @@ mono_assembly_get_image (MonoAssembly *assembly)
 MonoAssemblyName *
 mono_assembly_get_name (MonoAssembly *assembly)
 {
-	return &assembly->aname;
+	MonoAssemblyName *res;
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply mono/metadata/assembly.c]: Not really. I should probably drop this one.

[the remark this answers is not in the document]
Comment (jaykrell) [reply mono/metadata/assembly.c]: Maybe a comment, in case someone adds function calls? Or maybe cheap enough and more maintainable with the change?

**Comment by lambdageek** [comment]
@monojenkins build failed

**Comment by lambdageek** [comment]
Force-pushed to fix (hopefully) the toggleref failure:

* When SGen is configured for toggleref testing, it will look for an `Int32` field called `__test` in registered toggleref objects just prior to stopping the world.
* It does this using `mono_field_get_value`.
* In this PR, we made `mono_field_get_value` perform a transition to GC Unsafe mode.
* When that transition is necessary (under coop or hybrid), we copy a portion of the stack data
* That copied stack data ends up pinning a toggleref object since it was live when we did the transition.
* The tests expected that toggleref to be garbage and fail.

The fix is to add a `mono_field_get_value_internal` which requires the runtime to already be in GC Unsafe mode and doesn't do the thread state transition itself.  Then we call the `_internal` function from the toggleref test code.

**Pull request merged**